### ISSUE TITLE
Skip overide file if output file is near 0%

### ISSUE THREAD

**Issue by the-liquid-metal** [Lymphatus/caesium-image-compressor#235]
I have a bunch of files which i don't know whether it already compressed or not. Usually i (re)compressed those files to determine the status. This is laborious task and wasting time.

Cesium is able to calculate the percentage of the output file and determine whether the output file is larger than the source file or not. I use these parameters to determine whether my files have been compressed or not. A file is considered compressed if the percentage is close to 0%. I usually use 2% as a threshold.

It would be helpful if Cesium can set configurable threshold to skip override already compressed file. Moreover the info text **"skipped: below the threshold"** shold be shown.

**Comment by Lymphatus** [comment]
This is an interesting suggestion, even though I have other plans to address your specific issue.
Linked to #49 the idea is to build a local database of already compressed files, so Caesium knows if a file needs to be re-compressed or not.

You suggestion may still be useful in other scenarios, where - for example - you don't want to risk a loss of quality for just a small % of compression.

**Comment by the-liquid-metal** [comment]
> ... to build a local database of already compressed files...
> 

I am not a fan of external meta file:
- it can be deleted accidentally
- it can't be merged if we want to merge source folders manually

If there must be metadata, then it must be stored in that file too #237 

Some applications already perform compression when manipulating images, so cesium will always perform compression even if the file is already compressed efficiently if the meta file does not exist.

**Comment by anter69** [comment]
I'd love to see this idea implemented. There could be an extra box on the compression tab, like `do not compress if reduction is less than __%` (let's say 0..10%), and Caesium would simply skip writing the output in such cases. I believe it would be little work to implement, as there is already a size check to see if the output is greater than the input, basically that should be adjusted with the user defined percentage.

Keep up the great work!

**Comment by AntonKlink** [comment]
I believe JPEGmini adds some specific metadata to each file it has compressed and checks for the existence of such metadata even before starting the compression. If such metadata exists, it means JPEGmini has already compressed the file in the past and thus it skips compressing the file. This solution would make it universally applicable anywhere, enabling skipping the compression of already compressed files on different computers etc, wheres an external database would only work on one specific computer but make it almost impossible to achieve the same result on another computer. Also it would save CPU cycles, as the compression is completely skipped instead of going through compression and then comparing the output sizes. Wouldn't that be the best solution? It can also be a settings option "skip compression of files already compressed by Caeasium in the past".

**Comment by Lymphatus** [comment]
I see a couple of (minor) issues to write metadata to compressed images that comes to mind:
- It adds a (tiny) amount of size to each image, which is not desired
- Not all image formats handle metadata as JPEG does, so each format must have its own implementation
- As @anter69 said, just putting a flag that tells if it was already compressed it's not enough, because you might want to try again and compress further. A solution would be to store the compression settings (not only quality, but other options that could impact size too) in the metadata.
- Maybe more but it does not come to my mind now

I liked the database a bit more because it's a bit more flexible than metadata. For the installed version it would be stored as a file somewhere into the AppData folders, but for the portable it could be in the same directory as the executable, making it portable.

> an external database would only work on one specific computer but make it almost impossible to achieve the same result on another computer

This can be true, if the database uses as a key to identify the image something like the path of the image in the current computer. But if, for example, the key is the hash of the image content, it could be portable.

**Comment by anter69** [comment]
I think the quickest and easiest solution would be to use the already existing size check after compression, as explained above.

**Comment by sirtonyx** [comment]
Yes, please, no databases but instead a tag in the image file itself. I move photos between five different computers and a database on one computer would make this system completely useless on the other four computers. Even if working on only one computer with all your images, a database system is just one step away from being completely useless with a simple crash of the operating system and I need to reinstall or upgrade the OS or the computer/laptop itself. Would it really be so hard just to add a simple tag into the image file itself just like JPEG mini does?
